### PR TITLE
CMR-5106 Changing Entry Title of a collection causes it to disappear from searches until the ACL is re-created

### PR DESCRIPTION
**Pull request by daniel-zamora**
This fixes granule searching, the prior PR only fixed collection searches.

### CODE REVIEW

File: search-app/src/cmr/search/services/acls/granule_acls.clj [hunk 1]
```diff
@@ -152,22 +152,25 @@
...
                                       q/->CollectionQueryCondition)
             temporal-cond (some-> temporal temporal->query-condition q/->CollectionQueryCondition)]
-
         ;; If there's no entry title condition the other conditions must be scoped by provider
-        (gc/and-conds (remove nil? [(or entry-titles-cond colls-in-prov-cond)
+        ;; In anycase we want to use concept ids instead of entry-titles, because entry-titles can and
+        ;; do frequently change, so to prevent temporary loss of permissions concept ids are used
+        ;; when present.  There really shouldn't be a case where entry-tiles exist and concept-ids do
+        ;; not.
+        (gc/and-conds (remove nil? [(or concept-ids-cond entry-titles-cond)
```
Comment (ygliuvt) [inline search-app/src/cmr/search/services/acls/granule_acls.clj]: The feels wrong. We are dropping the `colls-in-prov-cond` which is constructed from the query params from the search. Why?

File: search-app/src/cmr/search/services/acls/granule_acls.clj [hunk 2]
```diff
@@ -152,22 +152,25 @@
...
-
 (defn collection-identifier->query-condition
   "Converts an acl collection identifier to an query condition. Switches implementations based
   on whether the user's query contained collection ids. This implementation assumes the query-coll-ids
   passed in are limited to the provider of the collection identifier."
   [context query-coll-ids provider-id collection-identifier]
   (let [colls-in-prov-cond (provider->collection-condition query-coll-ids provider-id)]
-    (if-let [{:keys [entry-titles access-value temporal]} collection-identifier]
+    (if-let [{:keys [concept-ids entry-titles access-value temporal]} collection-identifier]
       (let [entry-titles-cond (provider+entry-titles->collection-condition
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/acls/granule_acls.clj]: Why are we keeping the entry-titles-cond? Shouldn't every ACL now have concept-ids in them? I thought that was something that was done with the ECHO-ACL/CMR ACL sync functionality.

Is it still possible that an ACL could be saved with the concept-ids?

File: search-app/src/cmr/search/services/acls/granule_acls.clj [hunk 3]
```diff
@@ -152,22 +152,25 @@
...
+            concept-ids-cond (provider->collection-condition concept-ids provider-id)
             access-value-cond (some-> (access-value->query-condition access-value)
                                       q/->CollectionQueryCondition)
             temporal-cond (some-> temporal temporal->query-condition q/->CollectionQueryCondition)]
-
         ;; If there's no entry title condition the other conditions must be scoped by provider
-        (gc/and-conds (remove nil? [(or entry-titles-cond colls-in-prov-cond)
+        ;; In anycase we want to use concept ids instead of entry-titles, because entry-titles can and
+        ;; do frequently change, so to prevent temporary loss of permissions concept ids are used
+        ;; when present.  There really shouldn't be a case where entry-tiles exist and concept-ids do
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/acls/granule_acls.clj]: entry-titles typo

File: search-app/src/cmr/search/services/acls/granule_acls.clj [hunk 4]
```diff
@@ -152,22 +152,25 @@
...
-    (if-let [{:keys [entry-titles access-value temporal]} collection-identifier]
+    (if-let [{:keys [concept-ids entry-titles access-value temporal]} collection-identifier]
       (let [entry-titles-cond (provider+entry-titles->collection-condition
                                 context query-coll-ids provider-id entry-titles)
+            concept-ids-cond (provider->collection-condition concept-ids provider-id)
             access-value-cond (some-> (access-value->query-condition access-value)
                                       q/->CollectionQueryCondition)
             temporal-cond (some-> temporal temporal->query-condition q/->CollectionQueryCondition)]
-
         ;; If there's no entry title condition the other conditions must be scoped by provider
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/acls/granule_acls.clj]: I don't think this comment is valid anymore.

File: search-app/src/cmr/search/services/acls/granule_acls.clj [hunk 5]
```diff
@@ -152,22 +152,19 @@
...
-                                context query-coll-ids provider-id entry-titles)
+    (if-let [{:keys [concept-ids entry-titles access-value temporal]} collection-identifier]
+      (let [concept-ids-cond (provider->collection-condition concept-ids provider-id)
             access-value-cond (some-> (access-value->query-condition access-value)
                                       q/->CollectionQueryCondition)
             temporal-cond (some-> temporal temporal->query-condition q/->CollectionQueryCondition)]
 
-        ;; If there's no entry title condition the other conditions must be scoped by provider
-        (gc/and-conds (remove nil? [(or entry-titles-cond colls-in-prov-cond)
+        (gc/and-conds (remove nil? [(or concept-ids-cond colls-in-prov-cond)
```
Comment (ygliuvt) [inline search-app/src/cmr/search/services/acls/granule_acls.clj]: I don't think the logic here is correct. `provider->collection-condition` will just find all collections within the provider when the `concept-ids` is nil. That will always override the `colls-in-prov-cond` and that is not the desired behavior.

Comment (daniel-zamora) [reply search-app/src/cmr/search/services/acls/granule_acls.clj]:
```(defmulti provider->collection-condition
  "Converts a provider id from an ACL into a collection query condition that will find all collections
  in the provider. query-coll-ids should be a list of the collection ids given by the user in their
  original query (or collection ids resolved from collection query conditions) limited to the
  current provider."
  (fn [query-coll-ids provider-id]
    (if (empty? query-coll-ids)
      :no-query-coll-ids
      :with-query-coll-ids)))

(defmethod provider->collection-condition :no-query-coll-ids
  [query-coll-ids provider-id]
  (q/->CollectionQueryCondition
    (cqm/string-condition :provider-id provider-id true false)))

(defmethod provider->collection-condition :with-query-coll-ids
  [query-coll-ids provider-id]
  (cqm/string-conditions :collection-concept-id query-coll-ids true))
```

Comment (daniel-zamora) [reply search-app/src/cmr/search/services/acls/granule_acls.clj]: If concept-ids is nil it uses a provider-id condition.

Comment (daniel-zamora) [reply search-app/src/cmr/search/services/acls/granule_acls.clj]: Maybe I should remove the colls-in-prov-cond completely from that line.

Comment (ygliuvt) [reply search-app/src/cmr/search/services/acls/granule_acls.clj]: I think the correct logic is to AND these two collection conditions.

File: system-int-test/test/cmr/system_int_test/search/acls/granule.clj [hunk 6]
```diff
@@ -338,3 +338,47 @@
...
+(deftest granule-search-changed-entry-titles-test
+  (let [group1-concept-id (e/get-or-create-group (s/context) "group1")
+        guest-token (e/login-guest (s/context))
+        user1-token (e/login (s/context) "user1" [group1-concept-id])
+        user2-token (e/login (s/context) "user2")
+        coll1 (d/ingest "PROV1" (dc/collection {:entry-title "coll1"
+                                                :native-id "coll1"}))
+        coll2 (d/ingest "PROV1" (dc/collection {:entry-title "coll2"
+                                                :native-id "coll2"}))
+        coll1-edited (-> coll1
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/acls/granule.clj]: Remove `coll1-edited` as it is not used in the test.

File: system-int-test/test/cmr/system_int_test/search/acls/granule.clj [hunk 7]
```diff
@@ -338,3 +338,47 @@
...
+                                                   "PROV1" (e/coll-id ["coll1" "coll2"])))
+    (d/assert-refs-match [gran1 gran2 gran3 gran4] (search/find-refs
+                                                    :granule
+                                                    {:token user1-token
+                                                     :concept-id (map :concept-id all-colls)}))
+    (d/assert-refs-match [] (search/find-refs :granule {:token user2-token
+                                                        :concept-id (map :concept-id all-colls)}))
+    (d/assert-refs-match [] (search/find-refs :granule {:token guest-token
+                                                        :concept-id (map :concept-id all-colls)}))
+    (d/ingest "PROV1" (dc/collection {:entry-title "coll1-edited"
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/acls/granule.clj]: Add comment before this line saying something like "Update coll1 to a new entry-title" to make the test easier to follow.

File: system-int-test/test/cmr/system_int_test/search/acls/granule.clj [hunk 8]
```diff
@@ -338,3 +338,47 @@
...
+                                      :native-id "coll1"}))
+
+    (index/wait-until-indexed)
+    (dev-sys-util/clear-caches)
+
+    (d/assert-refs-match [gran1 gran2 gran3 gran4] (search/find-refs
```
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/acls/granule.clj]: Add a comment saying something like "Search as user1 still finds coll1's granules."

**Pull request merged**